### PR TITLE
Master search panel

### PR DESCRIPTION
**Pull request by Polymorphe57**
Link to task:  https://www.odoo.com/web#id=1892462&action=333&active_id=133&model=project.task&view_type=form&menu_id=4720

### CODE REVIEW

File: doc/reference/views.rst [hunk 1]
```diff
@@ -1106,6 +1106,43 @@ Possible children of the view element are:
        * kanban-specific CSS
        * kanban structures/widgets (vignette, details, ...)
 
+``searchpanel``
+  allows to display a search panel on the left of the kanban view.
+  This tool allows to quickly filter data on the basis of predefined
+  categories and filters.
+  The content of the search panel is determined by its chidren.
```
Comment (aab-odoo) [inline doc/reference/views.rst]: children

File: doc/reference/views.rst [hunk 2]
```diff
@@ -1106,6 +1106,43 @@ Possible children of the view element are:
...
+  The available child tags are:
+
+  ``category``
+  with a mandatory attribute ``name`` whose value is the name of a valid
+  (required) field of type many2one or selection.
+
+  Optional category attributes are:
+
+  * ``string``: determines the description of the category
```
Comment (aab-odoo) [inline doc/reference/views.rst]: title?

File: doc/reference/views.rst [hunk 3]
```diff
@@ -1106,6 +1106,43 @@ Possible children of the view element are:
...
+
+  * ``icon``: specifies which icon is used
```
Comment (aab-odoo) [inline doc/reference/views.rst]: specify default values for icon and color?

File: doc/reference/views.rst [hunk 4]
```diff
@@ -1106,6 +1106,43 @@ Possible children of the view element are:
...
+
+  * ``color``: determines the icon color
+
+  ``filter``
+  with a mandatory attribute ``name`` whose value is the name of a valid
+  field of type many2one, many2many, or selection.
+
+  Optional filter attributes are: ``string``, ``icon``, ``color``, and
+
+  * ``domain``: determines conditions that the comodel records has to satisfy. A domain might be used to express a dependency on a category. For instance
```
Comment (aab-odoo) [inline doc/reference/views.rst]: have to satisfy

File: doc/reference/views.rst [hunk 5]
```diff
@@ -1106,6 +1106,43 @@ Possible children of the view element are:
...
+
+  .. code-block:: xml
+
+    <searchpanel/>
+      <category name="department_id"/>
+      <filter name="manager_id" domain="[('department_id', 'parent_of', department_id)]"/>
+    <searchpanel/>
+
+  In the above example, the filter values available at screen will depend on
+  the value currently selected for the category department_id
```
Comment (aab-odoo) [inline doc/reference/views.rst]: not sure about the example

File: addons/web/controllers/kanban.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+from odoo.osv import expression
+
+CATEGORY_SUPPORTED_TYPES = ['many2one']
+FILTER_SUPPORTED_TYPES = ['many2one', 'many2many', 'selection']
+
+
+class KanbanSearchPanel(http.Controller):
+
+    @http.route('/web/kanban/get_search_panel_category', type='json', auth="user")
+    def get_search_panel_category(self, model, field_name):
```
Comment (aab-odoo) [inline addons/web/controllers/kanban.py]: add docstring

File: addons/web/controllers/kanban.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        parent_field = CoModel._fields[CoModel._parent_name]
+        if parent_field:
+            fields.append(parent_field.name)
+        return {
+            'parent_field': parent_field.name if parent_field else False,
+            'values': CoModel.search_read([], fields)
+        }
+
+    @http.route('/web/kanban/get_search_panel_filter', type='json', auth="user")
+    def get_search_panel_filter(self, model, field_name, **kwargs):
```
Comment (aab-odoo) [inline addons/web/controllers/kanban.py]: add docstring

File: addons/web/static/src/js/views/kanban/kanban_controller.js [hunk 8]
```diff
@@ -64,11 +83,27 @@ var KanbanController = BasicController.extend({
...
+     */
+    update: function (params, options) {
+        if (!this._searchPanel) {
+            return this._super.apply(this, arguments);
+        }
+        if (params.domain) {
+            this.controlPanelDomain = params.domain;
+        }
+        params.domain = this.controlPanelDomain.concat(this.searchPanelDomain);
+        return this._super(params, options).then(this._updateSearchPanel.bind(this));
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/kanban/kanban_controller.js]: two updates done sequentially

File: addons/web/static/src/js/views/kanban/search_panel.js [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+        var loadProm = this._fetchCategories().then(function () {
+            return self._fetchFilters();
+        });
+        return $.when(loadProm, this._super.apply(this, arguments));
+    },
+    /**
+     * @override
+     */
+    start: function () {
+        return $.when(this._render(), this._super.apply(this, arguments));
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/kanban/search_panel.js]: _render is sync

File: addons/web/static/src/js/views/kanban/search_panel.js [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+        Object.keys(this.categories).forEach(function (categoryId) {
+            var category = self.categories[categoryId];
+            evalContext[category.fieldName] = category.activeValueId;
+        });
+        var defs = Object.keys(this.filters).map(function (filterId) {
+            var filter = self.filters[filterId];
+            return self._rpc({
+                route: '/web/kanban/get_search_panel_filter',
+                params: {
+                    category_domain: self._getCategoryDomain(),
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/kanban/search_panel.js]: do this only once

File: addons/web/static/src/js/views/kanban/search_panel.js [hunk 11]
```diff
@@ -0,0 +1,513 @@
...
+                    comodel_domain: Domain.prototype.stringToArray(filter.domain, evalContext),
+                    disable_counters: filter.disableCounters,
+                    field_name: filter.fieldName,
+                    filter_domain: self._getFilterDomain(),
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/kanban/search_panel.js]: as well

File: addons/web/static/src/js/views/kanban/search_panel.js [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+            var category = self.categories[categoryId];
+            evalContext[category.fieldName] = category.activeValueId;
+        });
+        var defs = Object.keys(this.filters).map(function (filterId) {
+            var filter = self.filters[filterId];
+            return self._rpc({
+                route: '/web/kanban/get_search_panel_filter',
+                params: {
+                    category_domain: self._getCategoryDomain(),
+                    comodel_domain: Domain.prototype.stringToArray(filter.domain, evalContext),
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/kanban/search_panel.js]: same

File: addons/web/static/src/js/views/kanban/search_panel.js [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+     * @returns {jQuery}
+     */
+    _renderFilter: function (filter) {
+        var $filter = $(qweb.render('SearchPanel.Filter', {filter: filter}));
+
+        // set group inputs in intermediate state when necessary
+        Object.keys(filter.groups || {}).forEach(function (groupId) {
+            var state = filter.groups[groupId].state;
+            // group 'false' is not displayed
+            if (groupId !== 'false' && state === 'intermediate') {
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/kanban/search_panel.js]: 'false'?

File: addons/web/static/src/js/views/kanban/search_panel.js [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+        });
+
+        return $filter;
+    },
+    /**
+     * Compute the current searchPanel domain based on categories and filters,
+     * and notify environment of the domain change.
+     *
+     * Note that this assumes that the environment will update the searchPanel.
+     * FIXME: should we change that logic and update directly?
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/kanban/search_panel.js]: this is ok i think

File: doc/reference/views.rst (hunk 5 above)
Comment (aab-odoo) [reply doc/reference/views.rst]: also document disable_counters

File: doc/reference/views.rst (hunk 4 above)
Comment (aab-odoo) [reply doc/reference/views.rst]: overflow

File: addons/web/static/src/js/views/kanban/search_panel.js [hunk 15]
```diff
@@ -0,0 +1,513 @@
...
+        // unfold ancestor values of active value to make it is visible
+        if (category.activeValueId) {
+            var parentValueIds = this._getParentValueIds(category, category.activeValueId);
+            parentValueIds.forEach(function (parentValue) {
+                category.values[parentValue].folded = false;
+            });
+        }
+    },
+    /**
+     * @private
```
[the remark this answers is not in the document]
Comment (aab-odoo) [reply addons/web/static/src/js/views/kanban/search_panel.js]: don't reload on page change

File: addons/web/static/src/js/views/kanban/search_panel.js [hunk 16]
```diff
@@ -0,0 +1,526 @@
...
+                route: '/web/kanban/get_search_panel_filter',
+                params: {
+                    category_domain: categoryDomain,
+                    comodel_domain: Domain.prototype.stringToArray(filter.domain, evalContext),
+                    disable_counters: filter.disableCounters,
+                    field_name: filter.fieldName,
+                    filter_domain: filterDomain,
+                    group_by: filter.groupBy || false,
+                    model: self.model,
+                    search_domain: self.searchDomain,
```
Comment (rco-odoo) [inline addons/web/static/src/js/views/kanban/search_panel.js]: in alphabetical order, how can you read this?

File: addons/web/controllers/kanban.py (hunk 6 above)
Comment (rco-odoo) [reply addons/web/controllers/kanban.py]: changed into model method; comment is no longer relevant

File: addons/web/controllers/kanban.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+                        'group_id', 'group_sequence'.
+        }
+        """
+        comodel_domain = kwargs.get('comodel_domain', [])
+        disable_counters = kwargs.get('disable_counters', False)
+        group_by = kwargs.get('group_by', False)
+        model_domain = expression.AND([
+            kwargs.get('search_domain', []),
+            kwargs.get('category_domain', []),
+            kwargs.get('filter_domain', []),
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply addons/web/controllers/kanban.py]: Justified by crappy override in documents.

File: addons/web/controllers/kanban.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+            'parent_field': False or the field name specified by _parent_name
+                        on the comodel of the field 'field_name'.
+            'values': array of dictionaries containing some info on the records
+                        available on the comodel of the field 'field_name'.
+                        The search panel uses the keys: 'id', 'name', 'count',
+                        'group_id', 'group_sequence'.
+        }
+        """
+        comodel_domain = kwargs.get('comodel_domain', [])
+        disable_counters = kwargs.get('disable_counters', False)
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply addons/web/controllers/kanban.py]: Let's forget about it for this time.

File: addons/web/static/src/js/views/kanban/search_panel.js [hunk 19]
```diff
@@ -0,0 +1,524 @@
...
+            var category = self.categories[categoryId];
+            if (category.activeValueId) {
+                domain.push([category.fieldName, '=', category.activeValueId]);
+            }
+            return domain;
+        }
+        return Object.keys(this.categories).reduce(reducer, []);
+    },
+    /**
+     * Compute and return the domain based on the current checked filters.
```
[the remark this answers is not in the document]
Comment (Polymorphe57) [reply addons/web/static/src/js/views/kanban/search_panel.js]: done

File: addons/web/static/src/js/views/kanban/kanban_view.js [hunk 20]
```diff
@@ -100,14 +176,75 @@ var KanbanView = BasicView.extend({
...
+                id: sectionId,
+                index: index,
+                type: childNode.attrs.type,
+            };
+            if (section.type === 'category') {
+                section.color = section.color || '#875A7B';
+                section.icon = section.icon || 'fa-folder';
+            } else if (section.type === 'filter') {
+                section.color = section.color || '#D59244';
+                section.disableCounters = !!pyUtils.py_eval(childNode.attrs.disable_counters || '0');
```
[the remark this answers is not in the document]
Comment (Polymorphe57) [reply addons/web/static/src/js/views/kanban/kanban_view.js]: not done. not a problem

File: addons/web/models/models.py [hunk 21]
```diff
@@ -3,13 +3,20 @@
 import pytz
 
 from odoo import _, api, fields, models
+from odoo.osv.expression import AND
 from odoo.tools import lazy
+from odoo.exceptions import UserError
+
+CATEGORY_SUPPORTED_TYPES = ['many2one']
```
Comment (odony) [inline addons/web/models/models.py]: looks confusing at the top, maybe inline it in the methods, this way we don't have to find a proper name ;-)

File: addons/web/models/models.py [hunk 22]
```diff
@@ -114,3 +121,150 @@ def fields_view_get(self, view_id=None, view_type='form', toolbar=False, submenu
         if r['type'] == 'qweb':
             r['arch'] = '<qweb/>'
         return r
+
+    @api.model
+    def get_search_panel_category(self, field_name):
```
Comment (odony) [inline addons/web/models/models.py]:
```suggestion
    def search_panel_select_range(self, field_name):
```

File: addons/web/models/models.py [hunk 23]
```diff
@@ -114,3 +121,150 @@ def fields_view_get(self, view_id=None, view_type='form', toolbar=False, submenu
...
+        parent_name = Comodel._parent_name if Comodel._parent_name in Comodel._fields else False
+        if parent_name:
+            fields.append(parent_name)
+        return {
+            'parent_field': parent_name,
+            'values': Comodel.search_read([], fields),
+        }
+
+    @api.model
+    def get_search_panel_filter(self, field_name, **kwargs):
```
Comment (odony) [inline addons/web/models/models.py]:
```suggestion
    def search_panel_select_multi_range(self, field_name, **kwargs):
```

File: addons/web/models/models.py [hunk 24]
```diff
@@ -114,3 +121,150 @@ def fields_view_get(self, view_id=None, view_type='form', toolbar=False, submenu
...
+        :return: {
+            'parent_field': parent field on the comodel of field, or False
+            'values': array of dictionaries containing some info on the records
+                        available on the comodel of the field 'field_name'.
+                        The display name (and possibly parent_field) are fetched.
+        }
+        """
+        field = self._fields[field_name]
+        if field.type not in CATEGORY_SUPPORTED_TYPES:
+            raise UserError('Only types {} are supported for category (found type {})'.format(
```
Comment (odony) [inline addons/web/models/models.py]: translate + name placeholders with %s() that transifex seems to better support

```suggestion
            raise UserError(_('Only types %(supported_types)s are supported for category (found type %(field_type)s)') % (
```

File: addons/web/models/models.py [hunk 25]
```diff
@@ -114,3 +121,150 @@ def fields_view_get(self, view_id=None, view_type='form', toolbar=False, submenu
...
+        :return: a list of possible values, each being a dict with keys
+            'id' (value),
+            'name' (value label),
+            'count' (how many records with that value),
+            'group_id' (value of group),
+            'group_name' (label of group).
+        """
+        field = self._fields[field_name]
+        if field.type not in FILTER_SUPPORTED_TYPES:
+            raise UserError('Only types {} are supported for filter (found type {})'.format(
```
Comment (odony) [inline addons/web/models/models.py]: same thing as above

File: addons/web/models/models.py [hunk 26]
```diff
@@ -114,3 +118,152 @@ def fields_view_get(self, view_id=None, view_type='form', toolbar=False, submenu
         if r['type'] == 'qweb':
             r['arch'] = '<qweb/>'
         return r
+
+    @api.model
+    def search_panel_select_range(self, field_name):
+        """
+        Return possible category values, and their parent field.
```
Comment (aab-odoo) [inline addons/web/models/models.py]: doc?

File: addons/web/models/models.py [hunk 27]
```diff
@@ -114,3 +118,152 @@ def fields_view_get(self, view_id=None, view_type='form', toolbar=False, submenu
...
+            fields.append(parent_name)
+        return {
+            'parent_field': parent_name,
+            'values': Comodel.search_read([], fields),
+        }
+
+    @api.model
+    def search_panel_select_multi_range(self, field_name, **kwargs):
+        """
+        Return values of filter field, possibly with counters and group.
```
Comment (aab-odoo) [inline addons/web/models/models.py]: doc?

File: addons/web/static/tests/views/kanban_mobile_tests.js [hunk 28]
```diff
@@ -151,16 +159,49 @@ QUnit.module('Views', {
...
+            model: 'partner',
+            data: this.data,
+            arch: '<kanban>' +
+                    '<templates><t t-name="kanban-box">' +
+                        '<div>' +
+                            '<field name="foo"/>' +
+                        '</div>' +
+                    '</t></templates>' +
+                    '<searchpanel>' +
+                        '<category name="product_id"/>' +
```
Comment (aab-odoo) [inline addons/web/static/tests/views/kanban_mobile_tests.js]: oups

**Comment by aab-odoo** [comment]
robodoo rebase-ff

**Comment by aab-odoo** [comment]
robodoo r+

**Pull request closed without merging**